safenode/src/network/mod.rs

```diff
@@ -0,0 +1,376 @@
+    }
+
+    /// Request the content of the given file from the given peer.
+    pub async fn request_chunk(&mut self, peer: PeerId, xor_name: XorName) -> Result<Vec<u8>> {
+        let (sender, receiver) = oneshot::channel();
+        self.sender
+            .send(CmdToSwarm::RequestChunk {
+                xor_name,
+                peer,
+                sender,
+            })
+            .await
+            .expect("Command receiver not to be dropped.");
+        receiver.await.expect("Sender not be dropped.")
+    }
+
+    /// Respond with the provided file content to the given request.
+    pub async fn respond_chunk(&mut self, file: Vec<u8>, channel: ResponseChannel<ChunkResponse>) {
+        self.sender
+            .send(CmdToSwarm::RespondChunk { file, channel })
+            .await
+            .expect("Command receiver not to be dropped.");
+    }
+}
+
+pub struct EventLoop {
+    swarm: Swarm<SafeNodeBehaviour>,
+    command_receiver: mpsc::Receiver<CmdToSwarm>,
+    event_sender: mpsc::Sender<Event>,
+    pending_dial: HashMap<PeerId, oneshot::Sender<Result<()>>>,
+    pending_start_providing: HashMap<QueryId, oneshot::Sender<()>>,
+    pending_get_providers: HashMap<QueryId, oneshot::Sender<HashSet<PeerId>>>,
+    pending_request_file: HashMap<RequestId, oneshot::Sender<Result<Vec<u8>>>>,
+}
+
+impl EventLoop {
+    /// Creates the network components, namely:
+    ///
+    /// - The network client to interact with the network layer from anywhere
+    ///   within your application.
+    ///
+    /// - The network event stream, e.g. for incoming requests.
+    ///
+    /// - The network task driving the network itself.
+    pub async fn new(// secret_key_seed: Option<u8>,
+    ) -> Result<(Client, impl Stream<Item = Event>, EventLoop)> {
+        // Create a random key for ourselves.
+        let keypair = identity::Keypair::generate_ed25519();
+        let local_peer_id = PeerId::from(keypair.public());
+
+        // QUIC configuration
+        let quic_config = libp2p_quic::Config::new(&keypair);
+        let transport = libp2p_quic::async_std::Transport::new(quic_config);
+        let transport = transport
+            .map(|(peer_id, muxer), _| (peer_id, StreamMuxerBox::new(muxer)))
+            .boxed();
+        // Create a Kademlia instance and connect to the network address.
+        // Create a swarm to manage peers and events.
+        let swarm = {
+            // Create a Kademlia behaviour.
+            let mut cfg = KademliaConfig::default();
+            cfg.set_query_timeout(Duration::from_secs(5 * 60));
+            let kademlia =
+                Kademlia::with_config(local_peer_id, MemoryStore::new(local_peer_id), cfg);
+            let mdns = mdns::async_io::Behaviour::new(mdns::Config::default(), local_peer_id)?;
+            let behaviour = SafeNodeBehaviour {
+                request_response: request_response::Behaviour::new(
+                    ChunkStorageCodec(),
```

Comment:
Shouldn't we have a codec and protocol which embeds all type of storage data types rather than specific for chunks?

Reply:
I've made changes to make the req/res generic now! :+1: